File: codeforces/757/D/D.cpp
#include <bits/stdc++.h>
using namespace std;

int N;

constexpr size_t MAX_A = 5000009;

vector<long long> num_dividing(vector<long long>& A) {
  vector<long long> ret(MAX_A, 0);
  vector<long long> cnt(MAX_A, 0);
  for (auto a : A) {
    assert(a > 0);
    cnt[a]++;
  }

  for (int i=2; i<MAX_A; i++) {
    for (int j=i; j<MAX_A; j+=i) {
      ret[i] += cnt[j];
    }
  }

  ret[1] = A.size();
  return ret;
}


vector<long long> nd;

long long ans[MAX_A];

int main() {
  memset(ans, 0, sizeof(ans));
  int N;
  vector<long long> arr;
  cin >> N;
  for (int i=0; i<N; i++) {
    long long a;
    cin >> a;
    arr.push_back(a);
  }

  nd = num_dividing(arr);

  for (long long i=1; i<MAX_A; i++) {
    for (long long j=i; j<MAX_A; j+=i) {
      if (i == j) continue;
      // i -> j
      ans[j] = max(ans[j], ans[i] + i*(nd[i] - nd[j]));
    }
  }

  long long ret = -1;

  for (long long a : arr) {
    ret = max(ret, ans[a] + a*nd[a]);
  }
  cout << ret << endl;

  return 0;
}

File: codeforces/757/D/D.py
#!/usr/bin/env pypy3

from sys import stdin, stdout

def input(): return stdin.readline().strip()
def read_int_list(): return list(map(int, input().split()))
def read_int_tuple(): return tuple(map(int, input().split()))
def read_int(): return int(input())

### CODE HERE

from itertools import permutations
from math import gcd

def num_dividing_slow(A):
    MAX_A = max(A) + 1
    ret = [0]*MAX_A
    for a in A:
        for d in range(1, a+1):
            if a % d == 0:
                ret[d] += 1
    return ret

def num_dividing(A):
    # for each g < max(A), return | { a \in A : g | a }
    MAX_A = max(A) + 1
    ret = [0]*MAX_A
    cnt = [0]*MAX_A
    for a in A:
        cnt[a] += 1

    for i in range(2, MAX_A):
        for j in range(i, MAX_A, i):
            ret[i] += cnt[j]

    ret[1] = len(A)

    return ret

if False:
    tc = [9, 9]
    print(num_dividing_slow(tc))
    print(num_dividing(tc))
elif True:
    import random
    for N in range(1, 100):
        for _ in range(1000000):
            tc = [random.randint(1, 100) for _ in range(N)]
            if not (num_dividing(tc) == num_dividing_slow(tc)):
                print(tc)


def ans(A):
    N = len(A)
    MAX_A = max(A)+1
    cnt = [0]*MAX_A
    best_inc = [0]*MAX_A

    for a in A:
        cnt[a] += 1

    for i in range(MAX_A):
        best_inc[i] = cnt[i]*(i-1)

    for i in range(2, MAX_A):
        for j in range(2*i, MAX_A, i):
        #   // update i->j
            best_inc[j] = max(best_inc[j], cnt[j]*(j-1) + best_inc[i])

    return max(best_inc) + N

def score(arr):
    ret = 0
    curr = arr[0]
    ret += curr
    for a in arr[1:]:
        curr = gcd(curr, a)
        ret += curr
    return ret

def ans_slow(A):
    return max(score(p) for p in permutations(A))

if False:
    tc = [6, 1, 4]
    print(ans_slow(tc))
    print(ans(tc))

elif False:
    import random
    for _ in range(100):
        tc = [random.randint(1, 6) for _ in range(3)]
        if not (ans(tc) == ans_slow(tc)):
            print(tc)

# input()
# A = read_int_list()
# print(ans_slow(A))